lib/posix/clock.c

```diff
@@ -27,5 +47,74 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 	ts->tv_nsec = (s32_t) ((elapsed_msecs % MSEC_PER_SEC) *
 					USEC_PER_MSEC * NSEC_PER_USEC);
 
+	ts->tv_sec += base.tv_sec;
+	ts->tv_nsec += base.tv_nsec;
+	if (ts->tv_nsec > NSEC_PER_SEC) {
+		ts->tv_sec++;
+		ts->tv_nsec -= NSEC_PER_SEC;
+	}
+
 	return 0;
 }
+
+/**
+ * @brief Set the time of the specified clock.
+ *
+ * See IEEE 1003.1.
+ *
+ * Note that only the `CLOCK_REALTIME` clock can be set using this
+ * call.
+ */
+int clock_settime(clockid_t clock_id, const struct timespec *tp)
+{
+	struct timespec base;
+	int res;
+
+	if (clock_id != CLOCK_REALTIME) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	res = clock_gettime(clock_id, &base);
+	if (res != 0) {
+		return res;
+	}
+
+	base.tv_sec = tp->tv_sec - base.tv_sec;
+	base.tv_nsec = tp->tv_nsec - base.tv_nsec;
+
+	/* Make the tv_nsec field positive.  Only try a little,
+	 * assuming that for both clocks, the value is in a reasonable
+	 * range. */
+	if (base.tv_nsec < 0) {
+		base.tv_sec--;
+		base.tv_nsec += NSEC_PER_SEC;
+	}
+
+	rt_clock_base = base;
+
+	return 0;
+}
+
+/**
+ * @brief Get current real time.
+ *
+ * See IEEE 1003.1
+ */
+int gettimeofday(struct timeval *tv, const void *tz)
+{
+	struct timespec ts;
+	int res;
+
+	/* As per POSIX, "if tzp is not a null pointer, the behavior
+	 * is unspecified."  "tzp" is the "tz" parameter above. */
+	ARG_UNUSED(tv);
+
+	res = clock_gettime(CLOCK_REALTIME, &ts);
+	if (res == 0) {
```

Comment:
You don't really need this check.  There's nothing wrong with writing into the output parameter on error.

Reply:
I suppose, this just feels uncomfortable to me. POSIX doesn't really say anything here, because gettimeofday isn't allowed to return an error on POSIX (and we never will anyway).